src/dscanner/analysis/unused_result.d

```diff
@@ -24,174 +16,186 @@ import std.range : retro;
+	mixin VisitInstructionBlock!(AST.SwitchStatement);
+	mixin VisitInstructionBlock!(AST.SynchronizedStatement);
+	mixin VisitInstructionBlock!(AST.WithStatement);
+	mixin VisitInstructionBlock!(AST.TryCatchStatement);
+	mixin VisitInstructionBlock!(AST.TryFinallyStatement);
+
+	override void visit(AST.CompoundStatement compoundStatement)
+	{
+		foreach (statement; *compoundStatement.statements)
+		{
+			if (hasUnusedResult(statement))
+			{
+				auto lineNum = cast(ulong) statement.loc.linnum;
+				auto charNum = cast(ulong) statement.loc.charnum;
+				addErrorMessage(lineNum, charNum, KEY, MSG);
+			}
+
+			statement.accept(this);
+		}
+	}
+
+	override void visit(AST.IfStatement ifStatement)
+	{
+		if (hasUnusedResult(ifStatement.ifbody))
+		{
+			auto lineNum = cast(ulong) ifStatement.ifbody.loc.linnum;
+			auto charNum = cast(ulong) ifStatement.ifbody.loc.charnum;
+			addErrorMessage(lineNum, charNum, KEY, MSG);
+		}
+
+		if (ifStatement.elsebody && hasUnusedResult(ifStatement.elsebody))
+		{
+			auto lineNum = cast(ulong) ifStatement.elsebody.loc.linnum;
+			auto charNum = cast(ulong) ifStatement.elsebody.loc.charnum;
+			addErrorMessage(lineNum, charNum, KEY, MSG);
+		}
+
+		super.visit(ifStatement);
+	}
+
+	private bool hasUnusedResult(AST.Statement statement)
+	{
+		import dmd.astenums : TY;
+
+		auto exprStatement = statement.isExpStatement();
+		if (exprStatement is null)
+			return false;
+
+		auto callExpr = exprStatement.exp.isCallExp();
+		if (callExpr is null || callExpr.f is null)
+			return false;
+
+		auto typeFunction = callExpr.f.type.isTypeFunction();
+		if (typeFunction is null || typeFunction.next.ty == TY.Tvoid)
+			return false;
+
+		return true;
+	}
+
+	private template VisitInstructionBlock(T)
+	{
+		override void visit(T statement)
+		{
+			if (hasUnusedResult(statement._body))
+			{
+				auto lineNum = cast(ulong) statement._body.loc.linnum;
+				auto charNum = cast(ulong) statement._body.loc.charnum;
+				addErrorMessage(lineNum, charNum, KEY, MSG);
+			}
+
+			super.visit(statement);
+		}
+	}
 }
 
 unittest
 {
-    import dscanner.analysis.config : StaticAnalysisConfig, Check, disabledConfig;
-    import dscanner.analysis.helpers : assertAnalyzerWarnings;
-    import std.stdio : stderr;
-    import std.format : format;
+	import dscanner.analysis.config : StaticAnalysisConfig, Check, disabledConfig;
+	import dscanner.analysis.helpers : assertAnalyzerWarningsDMD;
+	import std.stdio : stderr;
+	import std.format : format;
 
-    StaticAnalysisConfig sac = disabledConfig();
-    sac.unused_result = Check.enabled;
+	enum string MSG = "Function return value is discarded";
+	StaticAnalysisConfig sac = disabledConfig();
+	sac.unused_result = Check.enabled;
 
-    assertAnalyzerWarnings(q{
-        void fun() {}
+	assertAnalyzerWarningsDMD(q{
+        int fun() { return 1; }
         void main()
         {
-            fun();
+            fun(); // [warn]: %s
         }
-    }c, sac);
+    }c.format(MSG), sac, 1);
```

Comment:
I would also leave the `void func` test just to make sure that we do not have any false positives.